README.md
# twitter-error-handler

[![npm version](https://badge.fury.io/js/twitter-error-handler.svg)](https://badge.fury.io/js/twitter-error-handler)
[![npm](https://img.shields.io/npm/dt/twitter-error-handler)](https://www.npmjs.com/package/twitter-error-handler)

Wrap errors from Twitter API calls like a boss. 😎

## Okay, what's this?

Let's take a look. You're writing a service where you make calls to the Twitter API. Suppose you're using the
package [Twit](https://github.com/ttezel/twit#readme). Your code would look something like this:

```javascript
return t.get('statuses/mentions_timeline', options)
    .then(r => r.data)
    .then(tweets => tweets.map(processTweet));
```

But how do you handle errors from the Twitter API? Maybe your API token is wrong, or your app has been suspended, or
you've hit a rate limit. Well, here's one way:

```javascript
return t.get('statuses/mentions_timeline', options)
    .catch(e => {
        const error = JSON.stringify(e);
        if (error.includes("Invalid or expired token")) {
            // uh-oh, your token is wrong, do stuff
        } else if (error.includes("Invalid / suspended application")) {
            // This is very bad, as it means your app is broken for its users
        } else if (error.includes("Rate limit exceeded")
            || error.includes("User is over daily status update limit")) {
            // implement a backoff so Twitter doesn't suspend your app
        } // you get the idea...
    })
    .then(r => r.data)
    .then(tweets => tweets.map(processTweet));
```

But this isn't very optimal, is it? What happens if Twitter changes their error messages? Also, that code makes my eyes
bleeeed. Here's what this package lets you do:

```javascript
const {wrapTwitterErrors, errors, codes} = require('twitter-error-handler');
return t.get('statuses/mentions_timeline', options)
    .catch(e => wrapTwitterErrors(e, 'statuses/mentions_timeline'))
    .catch(e => {
        if (e instanceof errors.ProblemWithAuth) {
            // uh-oh, your token is wrong, do stuff
        } else if (e instanceof errors.ProblemWithAppOrAccount) {
            // you need to take action
        } else if (e instanceof errors.RateLimited) {
            // implement a backoff so Twitter doesn't suspend your app
        }
    })
    .then(r => r.data)
    .then(tweets => tweets.map(processTweet));
```

If you mix in the [aargh package](https://github.com/shalvah/aargh), even neater:

```javascript
return t.get('statuses/mentions_timeline', options)
    .catch(e => wrapTwitterErrors(e, 'statuses/mentions_timeline'))
    .catch(e => {
        return aargh(e)
            .type(ProblemWithAuth, (e) => {
                // uh-oh, your token is wrong, do stuff
            })
            .type(ProblemWithAppOrAccount, notifyMe)
            .type(RateLimited, backOff)
            .throw();
    })
    .then(r => r.data)
    .then(tweets => tweets.map(processTweet));
```

## Installation

```
npm i twitter-error-handler
```

## Usage

This package wraps the errors into a bunch of classes that represent the kinds of errors you can get from Twitter. To
use this package, call the `wrapTwitterErrors` function with an `endpoint` and a `response` object within your first
catch (or try/catch or however you handle errors) statement after your Twitter API call.

- The `endpoint` is the path to the API you called. It's helpful so you can easily track what calls triggered what
  errors in your logs.

- The `response` object is the original Twitter API response, as a JavaScript object. This package was designed to work
  with [Twit](https://github.com/ttezel/twit#readme) (a very good Twitter API client). With Twit, the `response` object
  you pass in is the error in the catch block (as shown in the examples above). To use any other Twitter API client, you
  only need to pass in an `endpoint` and a `response` object matching the description above.

## Output

When you call the `wrapTwitterErrors` function within a catch block, it wil perform checks on any errors passed to it
and wrap the errors into a special Error object, and then throw it. The idea is to represent
all [possible Twitter API errors](https://developer.twitter.com/en/support/twitter-api/error-troubleshooting) as an
instance of one of the following:

- `ProblemWithYourAppOrAccount`: This covers cases where your Twitter app or user account is having issues (for
  instance, account locked or app suspended by Twitter). You'll probably want to take action ASAP. <details><summary>Covers the following errors:</summary>
  <ul>
  <li><code>ACCOUNT_LOCKED_TEMPORARILY</code> (326)</li>
  <li><code>ACCOUNT_SUSPENDED</code> (64)</li>
  <li><code>APP_SUSPENDED</code> (416)</li>
  <li><code>APP_MUZZLED</code> (261)</li>
  <li><code>CALLBACK_URL_NOT_APPROVED</code> (415)</li>
  <li><code>DESKTOP_APPLICATIONS_ONLY_SUPPORT_OOB_OAUTH</code> (417)</li>
  <li><code>CLIENT_NOT_PERMITTED_TO_DO_THAT</code> (87)</li>
  <li><code>APP_NOT_ALLOWED_TO_ACCESS_DMS</code> (93)</li>
  <li><code>YOUR_CREDENTIALS_DONT_COVER_THIS</code> (220)</li>
  <li><code>WE_FELT_LIKE_FLAGGING_YOU</code> (226)</li>
  </ul>
  </details>
- `RateLimited`: For errors caused by hitting a rate limit. You should implement backoff to avoid being banned by
  Twitter. Covers:
    - `RATE_LIMIT_EXCEEDED` (88)
    - `HIT_TWEET_LIMIT` (185)
    - `CANT_FOLLOW_MORE_PEOPLE_NOW_SO_CHILL` (161)
    - `REACHED_LIMIT_FOR_SPAM_REPORTS` (205)
- `ProblemWithAuth`: For errors caused by authentication issues. Covers:
    - `INVALID_OR_EXPIRED_TOKEN` (89)
    - `AUTH_FAILED` (32)
    - `AUTH_FAILED_DUE_TO_TIMESTAMPS` (135)
    - `UNABLE_TO_VERIFY_CREDENTIALS` (99)
    - `BAD_AUTH_DATA` (215)
- `ProblemWithTwitter`: For internal server errors and "Twitter is over capacity" errors. It's recommended to wait a few
  minutes before retrying. Covers:
  - `TWITTER_NEEDS_A_BREAK` (130)
  - `TWITTERS_DOWN_SORRY` (131)
- `ProblemWithPermissions`: For errors caused by permissions issues. For example, not allowed to view a protected tweet
  or DM a user. Covers:
    - `CANT_DM_THIS_USER` (150)
    - `COULDNT_DM` (151)
    - `TWEET_PROTECTED` (179)
    - `REPLIES_RESTRICTED` (433)
    - `BLOCKED_BY_USER` (136)
    - `OWNER_MUST_ALLOW_DMS_FROM_ANYONE` (214)
    - `CANT_DM_THE_FELLA` (349)
    - `TWEET_RESTRICTED_BY_TWITTER` (425)
- `BadRequest`: For errors caused by bad requests (invalid parameters, overly long values, invalid urls etc)
  . <details><summary>Covers the following errors:</summary>
  <ul>
  <li><code>INVALID_COORDINATES</code> (3)</li>
  <li><code>NO_DATA_FOR_COORDINATES_AND_RADIUS</code> (7)</li>
  <li><code>NO_DATA_FOR_THAT_ID</code> (8)</li>
  <li><code>MUST_PROVIDE_VALID_GEO_DATA</code> (12)</li>
  <li><code>CANT_REPORT_YOURSELF_FOR_SPAM</code> (36)</li>
  <li><code>REPLIED_TO_UNAVAILABLE_TWEET</code> (385)</li>
  <li><code>MISSING_QUERY_PARAMS</code> (25)</li>
  <li><code>PARAMETER_MISSING</code> (38)</li>
  <li><code>ATTACHMENT_URL_INVALID</code> (44)</li>
  <li><code>TWEET_TOO_LONG</code> (186)</li>
  <li><code>DUPLICATE_TWEET</code> (187)</li>
  <li><code>INVALID_URL_PARAMETER</code> (195)</li>
  <li><code>ACCOUNT_UPDATE_FAILED</code> (120)</li>
  <li><code>ALREADY_LIKED_THAT_SHIT</code> (129)</li>
  <li><code>ALREADY_FOLLOWED</code> (160)</li>
  <li><code>USER_NOT_IN_THE_LIST</code> (110)</li>
  <li><code>COULDNT_DETERMINE_SOURCE_USER</code> (163)</li>
  <li><code>DEVICE_ERROR</code> (203)</li>
  <li><code>HAHA_CANT_MUTE_YOURSELF</code> (271)</li>
  <li><code>CANT_UNMUTE_BECAUSE_YOU_WERENT_MUTING</code> (272)</li>
  <li><code>GIFS_NOT_ALLOWED_WITH_OTHER_IMAGES</code> (323)</li>
  <li><code>MEDIA_ID_GOT_A_PROBLEM</code> (324)</li>
  <li><code>DIDNT_FIND_A_MEDIA_ID</code> (325)</li>
  <li><code>ALREADY_RETWEETED</code> (327)</li>
  <li><code>DM_TOO_LONG</code> (354)</li>
  <li><code>SUBSCRIPTION_ALREADY_EXISTS</code> (355)</li>
  <li><code>ONLY_ONE_ATTACHMENT_TYPE_ALLOWED</code> (386)</li>
  <li><code>YOU_NEED_TO_ENABLE_COOKIES</code> (404)</li>
  <li><code>INVALID_URL</code> (407)</li>
  <li><code>INVALID_CONVERSATION_CONTROLS</code> (431)</li>
  </ul>
  </details>
- `NotFound`: For when the requested resource (eg Tweet/user) wasn't found. <details><summary>Covers the following
  errors:</summary>
  <ul>
  <li><code>NO_LOCATION_FOR_THAT_IP_ADDRESS</code> (13)  </li><li><code>NO_USER_MATCHES_THOSE_TERMS</code> (17)  </li><li><code>NOT_FOUND</code> (34)  </li><li><code>COULDNT_FIND_USER</code> (50)  <li><code>USER_NOT_FOUND_IN_THE_LIST</code> (109)  <li><code>NO_SUCH_TWEET</code> (144)  <li><code>USER_SUSPENDED</code> (63)  </li><li><code>TWEET_UNAVAILABLE</code> (422)  <li><code>TWEET_UNAVAILABLE_FOR_VIOLATING_RULES</code> (421)</li>
  </ul>
  </details>
- `TwitterApiError`: the base error class. Used for any errors which don't fall into the above groups; for instance,
  using a retired endpoint or no SSL. Currently covers:
  - `GOTO_NEW_API` (68)
  - `SSL_REQUIRED` (92)
  - `THIS_ENDPOINT_HAS_BEEN_RETIRED` (251)

> I've also included some error codes not mentioned in the docs (as at the time of writing), but which I've personally received.

If you need to take action based on a specific type of error, I recommend checking the response code rather than relying
on the inferred class.

Each error thrown by this package has the following properties:

- `code`: The error code from the Twitter response. You can programmatically inspect the code and decide what specific
  action to take. For instance, for a rate limit error, to back off for a varying number of minutes depending on the
  error, we could do (from the last example):

```javascript
     return aargh(e)
    .type(RateLimited, (e) => {
        if (e.code === codes.CANT_FOLLOW_MORE_PEOPLE_NOW_SO_CHILL) {
            retryIn({hours: 24});
        } else if (e.code === codes.HIT_TWEET_LIMIT) {
            retryIn({minutes: 30});
        }
    })
    .throw();
```

The various error codes are provided as properties of the `codes` object.

- `name`: The name of the error class
- `errors`: The original array of errors from the Twitter API response (as gotten from the `response` value passed to
  the `wrapTwitterErrors` function)
- `endpoint`: The endpoint for the API call (the same value passed to the `wrapTwitterErrors` function)

Calling the `valueOf()` method of any of the error objects returns a nice JSON representation of the above properties (
useful for logging purposes).

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres (or tries to🤷‍♂️) to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.1.0] - Monday, 8 February 2021
- Re-added `USER_MUST_VERIFY_LOGIN` (231).
- Added undocumented code `BLOCKED_BY_USER` (136)
- Moved `CANT_DM_THE_FELLA` (349) from `BadRequest` to `ProblemWithPermissions`

## [3.0.0] - Monday, 8 February 2021
- Added a new error type, `NotFound`
- Renamed some of the codes
  - `NO_LOCATION_ASSOCIATED_WITH_THE_SPECIFIED_IP_ADDRESS` -> `NO_LOCATION_FOR_THAT_IP_ADDRESS`,
  - `NO_USER_MATCHES_FOR_SPECIFIED_TERMS` -> `NO_USER_MATCHES_THOSE_TERMS`,
  - `ATTACHMENT__URL_INVALID` -> `ATTACHMENT_URL_INVALID`,
  - `CLIENT_IS_NOT_PERMITTED` -> `CLIENT_NOT_PERMITTED_TO_DO_THAT`
  - `PAST_ALLOWED_FIELD_LENGTH` -> `ACCOUNT_UPDATE_FAILED`
  - `THIS_ENDPOINT_HAS_BEEN_RETIRED_SO_FUCK_OFF` -> `THIS_ENDPOINT_HAS_BEEN_RETIRED`
- Removed `USER_MUST_VERIFY_LOGIN` (231), as it's no longer in the Twitter docs.
- Added the latest Twitter error codes,
  - `NO_DATA_FOR_COORDINATES_AND_RADIUS`: 7,
  - `NO_DATA_FOR_THAT_ID`: 8,
  - `MUST_PROVIDE_VALID_GEO_DATA`: 12,
  - `USER_NOT_FOUND_IN_THE_LIST`: 109,
  - `USER_NOT_IN_THE_LIST`: 110,
  - `COULDNT_DETERMINE_SOURCE_USER`: 163,
  - `DEVICE_ERROR`: 203,
  - `YOU_NEED_TO_ENABLE_COOKIES`: 404,
  - `TWEET_UNAVAILABLE`: 421,
  `TWEET_UNAVAILABLE_FOR_VIOLATING_RULES- `: 422,
  - `TWEET_RESTRICTED_BY_TWITTER`: 425,
  - `INVALID_CONVERSATION_CONTROLS`: 431

## [2.0.1] - Saturday, 21 September 2019
### Fixed
- Handle non-Twitter API errors properly (https://github.com/shalvah/twitter-error-handler/commit/f4a5c48c8d0d9275a003fd586e539b02073aec1f)

test.js
const tap = require('tap');

const {
    wrapTwitterErrors,
    errors: {
        ProblemWithAuth,
        NotFound
    },
    codes,
} = require('./index');

const Twit = require('twit');

const t = new Twit({
    consumer_key: process.env.TWITTER_CONSUMER_KEY || 'something',
    consumer_secret: process.env.TWITTER_CONSUMER_SECRET || 'something',
    access_token: process.env.TWITTER_ACCESS_TOKEN || 'something',
    access_token_secret: process.env.TWITTER_ACCESS_TOKEN_SECRE || 'something',
});

t.get('statuses/mentions_timeline')
    .catch(e => wrapTwitterErrors('statuses/mentions_timeline', e))
    .catch(e => {
        tap.true(e instanceof ProblemWithAuth);
        tap.same(e.code, codes.INVALID_OR_EXPIRED_TOKEN);
    });


t.get('nonexistentpath')
    .catch(e => wrapTwitterErrors('nonexistentpath', e))
    .catch(e => {
        tap.true(e instanceof NotFound);
        tap.same(e.code, codes.NOT_FOUND);
    });

// A simple mock. We want to assert other errors are also propagated.
t.get = () => Promise.reject(new RangeError());
t.get('statuses/mentions_timeline', )
    .catch(e => wrapTwitterErrors('statuses/mentions_timeline', e))
    .catch(e => {
        tap.true(e instanceof RangeError);
    });

const fakeResponse = {"message":"Invalid or expired token.","code":89,"allErrors":[{"code":89,"message":"Invalid or expired token."}],"twitterReply":{
    "errors":[{"code":89,"message":"Invalid or expired token."}]},"statusCode":401};


index.js
'use strict';

/**
 * @typedef {{code: number, message: string}|string} TwitError
 * @typedef {{allErrors: TwitError[], message?: string, code: number, twitterReply:  Object|string}} TwitResponse
 */

/**
 * Identity function. Coerces string/number literals to literal types.
 * @template {number} T
 * @param {T} v
 * @return {T}
 */
function c(v) {
    return v;
}

// See https://developer.twitter.com/en/docs/basics/response-codes.html
const codes = {
    INVALID_COORDINATES: c(3),
    NO_DATA_FOR_COORDINATES_AND_RADIUS: c(7),
    NO_DATA_FOR_THAT_ID: c(8),
    MUST_PROVIDE_VALID_GEO_DATA: c(12),
    NO_LOCATION_FOR_THAT_IP_ADDRESS: c(13),
    NO_USER_MATCHES_THOSE_TERMS: c(17),
    MISSING_QUERY_PARAMS: c(25),
    AUTH_FAILED: c(32),
    NOT_FOUND: c(34),
    CANT_REPORT_YOURSELF_FOR_SPAM: c(36),
    PARAMETER_MISSING: c(38),
    ATTACHMENT_URL_INVALID: c(44),
    COULDNT_FIND_USER: c(50),
    USER_SUSPENDED: c(63),
    ACCOUNT_SUSPENDED: c(64),
    GOTO_NEW_API: c(68),
    CLIENT_NOT_PERMITTED_TO_DO_THAT: c(87),
    RATE_LIMIT_EXCEEDED: c(88),
    INVALID_OR_EXPIRED_TOKEN: c(89),
    SSL_REQUIRED: c(92),
    APP_NOT_ALLOWED_TO_ACCESS_DMS: c(93),
    UNABLE_TO_VERIFY_CREDENTIALS: c(99),
    USER_NOT_FOUND_IN_THE_LIST: c(109),
    USER_NOT_IN_THE_LIST: c(110),
    ACCOUNT_UPDATE_FAILED: c(120),
    TWITTER_NEEDS_A_BREAK: c(130),
    TWITTERS_DOWN_SORRY: c(131),
    AUTH_FAILED_DUE_TO_TIMESTAMPS: c(135),
    BLOCKED_BY_USER: c(136), // Didn't find in docs, but I got it in a response as recently as last week🤷‍♀️
    ALREADY_LIKED_THAT_SHIT: c(139),
    NO_SUCH_TWEET: c(144),
    CANT_DM_THIS_USER: c(150),
    COULDNT_DM: c(151),
    ALREADY_FOLLOWED: c(160),
    CANT_FOLLOW_MORE_PEOPLE_NOW_SO_CHILL: c(161),
    COULDNT_DETERMINE_SOURCE_USER: c(163),
    TWEET_PROTECTED: c(179),
    HIT_TWEET_LIMIT: c(185),
    TWEET_TOO_LONG: c(186),
    DUPLICATE_TWEET: c(187),
    INVALID_URL_PARAMETER: c(195),
    DEVICE_ERROR: c(203),
    REACHED_LIMIT_FOR_SPAM_REPORTS: c(205),
    OWNER_MUST_ALLOW_DMS_FROM_ANYONE: c(214),
    BAD_AUTH_DATA: c(215),
    YOUR_CREDENTIALS_DONT_COVER_THIS: c(220),
    WE_FELT_LIKE_FLAGGING_YOU: c(226),
    USER_MUST_VERIFY_LOGIN: c(231),
    THIS_ENDPOINT_HAS_BEEN_RETIRED: c(251),
    APP_MUZZLED: c(261),
    HAHA_CANT_MUTE_YOURSELF: c(271),
    CANT_UNMUTE_BECAUSE_YOU_WERENT_MUTING: c(272),
    GIFS_NOT_ALLOWED_WITH_OTHER_IMAGES: c(323),
    MEDIA_ID_GOT_A_PROBLEM: c(324),
    DIDNT_FIND_A_MEDIA_ID: c(325),
    ACCOUNT_LOCKED_TEMPORARILY: c(326),
    ALREADY_RETWEETED: c(327),
    CANT_DM_THE_FELLA: c(349),
    DM_TOO_LONG: c(354),
    SUBSCRIPTION_ALREADY_EXISTS: c(355),
    REPLIED_TO_UNAVAILABLE_TWEET: c(385),
    ONLY_ONE_ATTACHMENT_TYPE_ALLOWED: c(386),
    YOU_NEED_TO_ENABLE_COOKIES: c(404),
    INVALID_URL: c(407),
    CALLBACK_URL_NOT_APPROVED: c(415),
    APP_SUSPENDED: c(416),
    DESKTOP_APPLICATIONS_ONLY_SUPPORT_OOB_OAUTH: c(417),
    TWEET_UNAVAILABLE: c(63),
    TWEET_UNAVAILABLE_FOR_VIOLATING_RULES: c(422),
    TWEET_RESTRICTED_BY_TWITTER: c(425),
    INVALID_CONVERSATION_CONTROLS: c(431),
    REPLIES_RESTRICTED: c(433),
};

class TwitterApiError extends Error {
    /**
     * @param {string} endpoint The endpoint which triggered the error
     * @param {TwitError[]} errors The errors object from the Twitter response
     * @param {string} type Category of the error
     *
     * Base error class covers:
     * - GOTO_NEW_API (68)
     * - SSL_REQUIRED (92)
     * - THIS_ENDPOINT_HAS_BEEN_RETIRED (251)
     */
    constructor(endpoint, errors, type = 'Unknown') {
        super(`[${type}]: Twitter API Error; endpoint=${endpoint}; error=${JSON.stringify(errors[0])}`);
        this.errors = errors;
        this.code = typeof errors[0] === "string" ? null : Number(errors[0].code);
        this.endpoint = endpoint;
        this.name = 'Twitter API Error';
    }

    valueOf() {
        return JSON.stringify({
            name: this.name,
            errors: this.errors,
            code: this.code,
            endpoint: this.endpoint,
        });
    }
}

/**
 * For when your Twitter app or user account is having issues (for instance, account locked or app suspended by Twitter).
 * You'll probably want to take action ASAP.
 * Covers:
 * - ACCOUNT_LOCKED_TEMPORARILY (326)
 * - ACCOUNT_SUSPENDED (64)
 * - APP_SUSPENDED (416)
 * - APP_MUZZLED (261)
 * - CALLBACK_URL_NOT_APPROVED (415)
 * - DESKTOP_APPLICATIONS_ONLY_SUPPORT_OOB_OAUTH (417)
 * - CLIENT_NOT_PERMITTED_TO_DO_THAT (87)
 * - APP_NOT_ALLOWED_TO_ACCESS_DMS (93)
 * - YOUR_CREDENTIALS_DONT_COVER_THIS (220)
 * - WE_FELT_LIKE_FLAGGING_YOU (226)
 */
class ProblemWithYourAppOrAccount extends TwitterApiError {
    constructor(endpoint, errors) {
        super(endpoint, errors, 'YourAppOrAccount');
        this.name = 'ProblemWithYourAppOrAccount';
    }
}

/**
 * For when the requested resource (eg Tweet/user) wasn't found.
 * Covers:
 * - NO_LOCATION_FOR_THAT_IP_ADDRESS (13)
 * - NO_USER_MATCHES_THOSE_TERMS (17)
 * - NOT_FOUND (34)
 * - COULDNT_FIND_USER (50)
 * - USER_NOT_FOUND_IN_THE_LIST (109)
 * - NO_SUCH_TWEET (144)
 * - USER_SUSPENDED (63)
 * - TWEET_UNAVAILABLE (422)
 * - TWEET_UNAVAILABLE_FOR_VIOLATING_RULES (421)
 */
class NotFound extends TwitterApiError {
    constructor(endpoint, errors) {
        super(endpoint, errors, 'NotFound');
        this.name = 'NotFound';
    }
}

/**
 * For internal server errors and "Twitter is over capacity" errors.
 * It's recommended to wait a few minutes before retrying.
 * Covers:
 * - TWITTER_NEEDS_A_BREAK (130)
 * - TWITTERS_DOWN_SORRY (131)
 */
class ProblemWithTwitter extends TwitterApiError {
    constructor(endpoint, errors) {
        super(endpoint, errors, 'Server');
        this.name = 'ProblemWithTwitter';
    }
}

/**
 * For errors caused by bad requests (invalid parameters, overly long values etc)
 * Covers:
 * - INVALID_COORDINATES (3)
 * - NO_DATA_FOR_COORDINATES_AND_RADIUS (7)
 * - NO_DATA_FOR_THAT_ID (8)
 * - MUST_PROVIDE_VALID_GEO_DATA (12)
 * - CANT_REPORT_YOURSELF_FOR_SPAM (36)
 * - REPLIED_TO_UNAVAILABLE_TWEET (385)
 * - MISSING_QUERY_PARAMS (25)
 * - PARAMETER_MISSING (38)
 * - ATTACHMENT_URL_INVALID (44)
 * - TWEET_TOO_LONG (186)
 * - DUPLICATE_TWEET (187)
 * - INVALID_URL_PARAMETER (195)
 * - ACCOUNT_UPDATE_FAILED (120)
 * - ALREADY_LIKED_THAT_SHIT (129)
 * - ALREADY_FOLLOWED (160)
 * - USER_NOT_IN_THE_LIST (110)
 * - COULDNT_DETERMINE_SOURCE_USER (163)
 * - DEVICE_ERROR (203)
 * - HAHA_CANT_MUTE_YOURSELF (271)
 * - CANT_UNMUTE_BECAUSE_YOU_WERENT_MUTING (272)
 * - GIFS_NOT_ALLOWED_WITH_OTHER_IMAGES (323)
 * - MEDIA_ID_GOT_A_PROBLEM (324)
 * - DIDNT_FIND_A_MEDIA_ID (325)
 * - ALREADY_RETWEETED (327)
 * - DM_TOO_LONG (354)
 * - SUBSCRIPTION_ALREADY_EXISTS (355)
 * - ONLY_ONE_ATTACHMENT_TYPE_ALLOWED (386)
 * - YOU_NEED_TO_ENABLE_COOKIES (404)
 * - INVALID_URL (407)
 * - INVALID_CONVERSATION_CONTROLS (431)
 */
class BadRequest extends TwitterApiError {
    constructor(endpoint, errors) {
        super(endpoint, errors, 'Client');
        this.name = 'BadRequest';
    }
}

/**
 * For errors caused by hitting a rate limit. You should implement backoff to avoid being banned by Twitter.
 * Covers:
 * - RATE_LIMIT_EXCEEDED (88)
 * - HIT_TWEET_LIMIT (185)
 * - CANT_FOLLOW_MORE_PEOPLE_NOW_SO_CHILL (161)
 * - REACHED_LIMIT_FOR_SPAM_REPORTS (205)
 */
class RateLimited extends TwitterApiError {
    constructor(endpoint, errors) {
        super(endpoint, errors, `RateLimited`);
        this.name = 'RateLimited';
    }
}

/**
 * For errors caused by authentication issues.
 * Covers:
 * - INVALID_OR_EXPIRED_TOKEN (89)
 * - AUTH_FAILED (32)
 * - AUTH_FAILED_DUE_TO_TIMESTAMPS (135)
 * - UNABLE_TO_VERIFY_CREDENTIALS (99)
 * - BAD_AUTH_DATA (215)
 */
class ProblemWithAuth extends TwitterApiError {
    constructor(endpoint, errors) {
        super(endpoint, errors, `Auth`);
        this.name = 'ProblemWithAuth';
    }
}

/**
 * For errors caused by permissions issues.
 * For example, not allowed to view a protected tweet or DM a user.
 * Covers:
 * - CANT_DM_THIS_USER (150)
 * - COULDNT_DM (151)
 * - TWEET_PROTECTED (179)
 * - REPLIES_RESTRICTED (433)
 * - BLOCKED_BY_USER (136)
 * - OWNER_MUST_ALLOW_DMS_FROM_ANYONE (214)
 * - CANT_DM_THE_FELLA (349)
 * - TWEET_RESTRICTED_BY_TWITTER (425)
 */
class ProblemWithPermissions extends TwitterApiError {
    constructor(endpoint, errors) {
        super(endpoint, errors, `Permissions`);
        this.name = 'ProblemWithPermissions';
    }
}

/**
 *
 * @param {string} endpoint
 * @param {TwitResponse} response
 */
const wrapTwitterErrors = (endpoint, response) => {
    const errors = response.allErrors;
    if (!errors || !errors.length) {
        // In case some other error happened, we handle that
        throw response;
    }

    if (typeof errors[0] === "string") {
        throw new TwitterApiError(endpoint, errors);
    }

    switch (errors[0].code) {
        case codes.RATE_LIMIT_EXCEEDED:
        case codes.HIT_TWEET_LIMIT:
        case codes.CANT_FOLLOW_MORE_PEOPLE_NOW_SO_CHILL:
        case codes.REACHED_LIMIT_FOR_SPAM_REPORTS:
            throw new RateLimited(endpoint, errors);
        case codes.TWITTER_NEEDS_A_BREAK:
        case codes.TWITTERS_DOWN_SORRY:
            throw new ProblemWithTwitter(endpoint, errors);
        case codes.ACCOUNT_LOCKED_TEMPORARILY:
        case codes.ACCOUNT_SUSPENDED:
        case codes.APP_SUSPENDED:
        case codes.APP_MUZZLED:
        case codes.CALLBACK_URL_NOT_APPROVED:
        case codes.DESKTOP_APPLICATIONS_ONLY_SUPPORT_OOB_OAUTH:
        case codes.CLIENT_NOT_PERMITTED_TO_DO_THAT:
        case codes.APP_NOT_ALLOWED_TO_ACCESS_DMS:
        case codes.YOUR_CREDENTIALS_DONT_COVER_THIS:
        case codes.WE_FELT_LIKE_FLAGGING_YOU:
            throw new ProblemWithYourAppOrAccount(endpoint, errors);
        case codes.INVALID_COORDINATES:
        case codes.NO_DATA_FOR_COORDINATES_AND_RADIUS:
        case codes.NO_DATA_FOR_THAT_ID:
        case codes.MUST_PROVIDE_VALID_GEO_DATA:
        case codes.CANT_REPORT_YOURSELF_FOR_SPAM:
        case codes.REPLIED_TO_UNAVAILABLE_TWEET:
        case codes.MISSING_QUERY_PARAMS:
        case codes.PARAMETER_MISSING:
        case codes.ATTACHMENT_URL_INVALID:
        case codes.TWEET_TOO_LONG:
        case codes.DUPLICATE_TWEET:
        case codes.INVALID_URL_PARAMETER:
        case codes.ACCOUNT_UPDATE_FAILED:
        case codes.ALREADY_LIKED_THAT_SHIT:
        case codes.ALREADY_FOLLOWED:
        case codes.USER_NOT_IN_THE_LIST:
        case codes.COULDNT_DETERMINE_SOURCE_USER:
        case codes.DEVICE_ERROR:
        case codes.HAHA_CANT_MUTE_YOURSELF:
        case codes.CANT_UNMUTE_BECAUSE_YOU_WERENT_MUTING:
        case codes.GIFS_NOT_ALLOWED_WITH_OTHER_IMAGES:
        case codes.MEDIA_ID_GOT_A_PROBLEM:
        case codes.DIDNT_FIND_A_MEDIA_ID:
        case codes.ALREADY_RETWEETED:
        case codes.DM_TOO_LONG:
        case codes.SUBSCRIPTION_ALREADY_EXISTS:
        case codes.ONLY_ONE_ATTACHMENT_TYPE_ALLOWED:
        case codes.YOU_NEED_TO_ENABLE_COOKIES:
        case codes.INVALID_URL:
        case codes.INVALID_CONVERSATION_CONTROLS:
            throw new BadRequest(endpoint, errors);
        case codes.NO_LOCATION_FOR_THAT_IP_ADDRESS:
        case codes.NO_USER_MATCHES_THOSE_TERMS:
        case codes.NOT_FOUND:
        case codes.COULDNT_FIND_USER:
        case codes.USER_NOT_FOUND_IN_THE_LIST:
        case codes.NO_SUCH_TWEET:
        case codes.USER_SUSPENDED:
        case codes.TWEET_UNAVAILABLE:
        case codes.TWEET_UNAVAILABLE_FOR_VIOLATING_RULES:
            throw new NotFound(endpoint, errors);
        case codes.INVALID_OR_EXPIRED_TOKEN:
        case codes.AUTH_FAILED:
        case codes.AUTH_FAILED_DUE_TO_TIMESTAMPS:
        case codes.UNABLE_TO_VERIFY_CREDENTIALS:
        case codes.BAD_AUTH_DATA:
            throw new ProblemWithAuth(endpoint, errors);
        case codes.CANT_DM_THIS_USER:
        case codes.COULDNT_DM:
        case codes.TWEET_PROTECTED:
        case codes.REPLIES_RESTRICTED:
        case codes.BLOCKED_BY_USER:
        case codes.OWNER_MUST_ALLOW_DMS_FROM_ANYONE:
        case codes.CANT_DM_THE_FELLA:
        case codes.TWEET_RESTRICTED_BY_TWITTER:
            throw new ProblemWithPermissions(endpoint, errors);
        case codes.GOTO_NEW_API:
        case codes.SSL_REQUIRED:
        case codes.THIS_ENDPOINT_HAS_BEEN_RETIRED:
        default:
            throw new TwitterApiError(endpoint, errors);
    }
};

module.exports = {
    codes,
    errors: {
        ProblemWithYourAppOrAccount,
        NotFound,
        ProblemWithPermissions,
        RateLimited,
        BadRequest,
        ProblemWithAuth,
        ProblemWithTwitter,
        TwitterApiError,
    },
    wrapTwitterErrors,
};